ArduCopter/baro_ground_effect.cpp

```diff
@@ -67,3 +71,18 @@ void Copter::update_ground_effect_detector(void)
     ahrs.setTakeoffExpected(gndeffect_state.takeoff_expected);
     ahrs.setTouchdownExpected(gndeffect_state.touchdown_expected);
 }
+
+// update ekf terrain height stable setting
+// when set to true, this allows the EKF to stabilize the normally barometer based altitude using a rangefinder
+// this is not related to terrain following
+void Copter::update_ekf_terrain_height_stable()
+{
+    // set to false if no position estimate
+    if (!position_ok() && !optflow_position_ok()) {
+        ahrs.set_terrain_hgt_stable(false);
+    }
+
+    // consider terrain height stable if vehicle is taking off or landing and moving less than 0.5 m/s
+    const bool is_stable = (flightmode->is_taking_off() || flightmode->is_landing()) && (ahrs.groundspeed() <= EKF_TERRAIN_HGT_STABLE_SPEED_MAX);
```

Comment:
I've looked at the code for is_taking_off() - it seems like it will almost never return true, except perhaps in auto modes. I personally think this is too restrictive.

Reply:
@andyp1per, yes, it's very restrictive.  It will only be true if the PILOT_TKOFF_ALT is set or the vehicle is doing an autonomous takeoff in Auto or Guided.